src/messaging/ReliableMessageManager.h

```diff
@@ -0,0 +1,126 @@
+#include <system/SystemTimer.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+namespace messaging {
+
+enum class SendMessageFlags : uint16_t;
+class ReliableMessageContext;
+
+class ReliableMessageManager
+{
+public:
+    /**
+     *  @class RetransTableEntry
+     *
+     *  @brief
+     *    This class is part of the CHIP Reliable Messaging Protocol and is used
+     *    to keep track of CHIP messages that have been sent and are expecting an
+     *    acknowledgment back. If the acknowledgment is not received within a
+     *    specific timeout, the message would be retransmitted from this table.
+     *
+     */
+    struct RetransTableEntry
+    {
+        RetransTableEntry();
+
+        ReliableMessageContext * rc;   /**< The context for the stored CHIP message. */
+        System::PacketBuffer * msgBuf; /**< A pointer to the PacketBuffer object holding the CHIP message. */
+        uint32_t msgId;                /**< The message identifier of the CHIP message awaiting acknowledgment. */
+        uint16_t msgSendFlags;
+        uint16_t nextRetransTimeTick; /**< A counter representing the next retransmission time for the message. */
+        uint8_t sendCount;            /**< A counter representing the number of times the message has been sent. */
+    };
+
+public:
+    ReliableMessageManager();
+    ~ReliableMessageManager();
+
+    void Init(chip::System::Layer & system) { mSystemLayer = &system; }
+    void Shutdown() {}
+
+    void FreeContext(ReliableMessageContext * rc);
+
+    uint64_t GetTickCounterFromTimePeriod(uint64_t period);
+    uint64_t GetTickCounterFromTimeDelta(uint64_t newTime);
+
+    void ExecuteActions();
+    void ProcessDelayedDeliveryMessage(ReliableMessageContext * rc, uint32_t PauseTimeMillis);
+    static void Timeout(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    CHIP_ERROR AddToRetransTable(ReliableMessageContext * rc, System::PacketBuffer * msgBuf, uint32_t messageId,
+                                 uint16_t msgSendFlags, RetransTableEntry ** rEntry);
+    void PauseRetransTable(ReliableMessageContext * rc, uint32_t PauseTimeMillis);
+    void ResumeRetransTable(ReliableMessageContext * rc);
+    bool CheckAndRemRetransTable(ReliableMessageContext * rc, uint32_t msgId);
+    CHIP_ERROR SendFromRetransTable(RetransTableEntry * entry);
+    void ClearRetransmitTable(ReliableMessageContext * rc);
+    void ClearRetransmitTable(RetransTableEntry & rEntry);
+    void FailRetransmitTableEntries(ReliableMessageContext * rc, CHIP_ERROR err);
+
+    void StartTimer();
+    void StopTimer();
+    void ExpireTicks();
+
+    // Functions for testing
+    int TestGetCountRetransTable();
+    void TestSetIntervalShift(uint16_t value) { mTimerIntervalShift = value; }
+
+public:
+    // public functions for ReliableMessageProtocol internal usage
+    CHIP_ERROR SendMessage(ReliableMessageContext * context, System::PacketBuffer * msgBuf, uint16_t sendFlags);
```

Comment:
Where is the implementation of SendMessage function?

Reply:
It will be integrate with `ExchangeContext`. We will find a way to associate `ReliableMessageContext` with `ExchangeContext` and here `SendMessage` will probably propagate to `SendMessage` in `ExchangeManager`.

Reply:
Thanks, I think this need to be implemented with SessionManager, just like what ExchangeContext does today. After CRMP is merged, ExchangeContext will switch to use CRMP to send message instead of using SessionManager.

Reply:
I think we can leave this implementation to another PR. Basically, CRMP is an internal module of messaging layer and ExchangeManager /ExchangeContext is its client. For TCP/BLE message, ExchangeContext use SecureSessionManager to send the message directly, for UDP message, ExchangeContext will use CRMP to send the message instead.